aten/src/ATen/native/cpu/PowKernel.cpp

```diff
@@ -35,62 +35,165 @@ void pow_tensor_tensor_kernel(TensorIterator& iter) {
-        cpu_kernel_vec(iter,
-          [=](scalar_t base) -> scalar_t {
-            return std::pow(base, exp);
-          },
-          [=](Vec base) -> Vec { return base.pow(exp); }
-        );
-      }
-    });
+    if ((iter.dtype() == ScalarType::Half)) {
+      // sqrt() & rsqrt() not used for exp = 0.5 & -0.5 respectively because
+      // their specialized variants for c10::Half do not exist in aten/cpu/vec256.
+      [&]() {
+        using scalar_t = decltype(c10::impl::ScalarTypeToCPPType<ScalarType::Half>::t);
+        auto exp = exp_scalar.to<scalar_t>();
+        using Vec = Vec256<scalar_t>;
+        if (exp == 2) {
+          cpu_kernel_vec(iter,
+            [](scalar_t base) -> scalar_t {
+              return base * base;
+            },
+            [](Vec base) -> Vec { return base * base; }
+          );
+        } else if (exp == 3) {
+          cpu_kernel_vec(iter,
+            [](scalar_t base) -> scalar_t {
+              return base * base * base;
+            },
+            [](Vec base) -> Vec { return base * base * base; }
+          );
+        } else if (exp == -1) {
+          cpu_kernel_vec(iter,
+            [](scalar_t base) -> scalar_t {
+              return 1.0 / base;
+            },
+            [](Vec base) -> Vec { return base.reciprocal(); }
+          );
+        } else if (exp == -2) {
+          cpu_kernel_vec(iter,
+            [](scalar_t base) -> scalar_t {
+              return 1.0 / (base * base);
+            },
+            [](Vec base) -> Vec { return (base * base).reciprocal(); }
+          );
+        } else {
+          cpu_kernel_vec(iter,
+            [=](scalar_t base) -> scalar_t {
+              return std::pow(base, exp);
+            },
+            [=](Vec base) -> Vec { return base.pow(exp); }
+          );
+        }
+      }();
+    } else if (iter.dtype() == ScalarType::BFloat16) {
+      [&]() {
+        using scalar_t = decltype(c10::impl::ScalarTypeToCPPType<ScalarType::BFloat16>::t);
+        auto exp = exp_scalar.to<scalar_t>();
+        using Vec = Vec256<scalar_t>;
+        if (exp == 0.5) {
+          cpu_kernel_vec(iter,
+            [=](scalar_t base) -> scalar_t {
+              return std::sqrt(float(base));
+            },
+            [](Vec base) -> Vec { return base.sqrt(); }
+          );
+        } else if (exp == 2) {
+          cpu_kernel_vec(iter,
+            [](scalar_t base) -> scalar_t {
+              return base * base;
+            },
+            [](Vec base) -> Vec { return base * base; }
+          );
+        } else if (exp == 3) {
+          cpu_kernel_vec(iter,
+            [](scalar_t base) -> scalar_t {
+              return base * base * base;
+            },
+            [](Vec base) -> Vec { return base * base * base; }
+          );
+        } else if (exp == -0.5) {
+          cpu_kernel_vec(iter,
+            [](scalar_t base) __ubsan_ignore_float_divide_by_zero__ -> scalar_t {
+              return 1.0 / std::sqrt(float(base));
```

Comment:
Just to be clear, `float()` here is an overloaded operator of `BFloat16`.

https://github.com/pytorch/pytorch/blob/f0e72e54cc3049e986adc5bff314dd0b649c780b/c10/util/BFloat16-inl.h#L19